### DISCUSSION TITLE
Language Specific Instructions

### DISCUSSION THREAD

**Discussion by yancyribbens** [universal-ctags/ctags#3604] [General]
Hello,

I'm curious if there is a curated list of suggestions as part of this repo for running ctags with a given language.  Recently I wrote a simple library for compiling ctags with Rust project dependencies https://crates.io/crates/run-ctags.  Is there a section of this repo that this could be added to?

Thanks!
-Yancy

**Comment by masatake** [comment]
I'm not good at English. So I'm not sure I understand what you wrote as you expected.

I understand that you are looking for web pages or documents of u-ctags where you can advertise your software.

https://docs.ctags.io/en/latest/other-projects.html This is one of the places.
The source code for the page is docs/other-project.rst.

Another one is `ctags-lang-rust.7.rst.`
There is no such document in our repository yet. However, I'm encouraging people to write a man page for each parser/indexer. Adding a new page for Rust and writing your software on the man page is a good idea. See https://docs.ctags.io/en/latest/man-pages.html.

The last ultimate one is implementing -r, semantic recursion option.
If `ctags -r Cargo.toml` can do the same as `run-ctags | ctags -R -L -`, people don't have to install your software.

**Comment by yancyribbens** [comment]
Hi @masatake,

If there is a way to implement semantic recursion using ctags without needing to install additional software, that would be best.  Could you point me to an example or give me some more information about how this could be accomplished?  Would this mean adding a PR to the `ctags` code base?

Cheers,
-Yancy

**Comment by masatake** [comment]
> Could you point me to an example or give me some more information about how this could be accomplished? Would this mean adding a PR to the ctags code base?

No real example. It is just a vague idea. 
In my design, so many parts are missed in ctags to implement such features.

One of the most important one is Toml parser (https://github.com/universal-ctags/ctags/pull/3509).
https://github.com/universal-ctags/ctags/issues/2428 is also related to this topic.

BTW ~The idea is just about Cargo.toml.~ The idea not is just about Cargo.toml.
foo.c:
```
#include <stdio.h>
int
main(void) { printf("hello, world\n"); }
```

`ctags -r foo.c` makes a tags file for foo.c, stdio.h and files included from stdio.h.

**Reply by yancyribbens** [reply]
Hi @masatake - parsing the Toml file of Cargo.toml is not enough.  The helper application I wrote invokes `cargo` the rust package manager and then parses the `json` produced by cargo to locate all libraries the program depends on (all dependencies).  I'm not sure this is something that can be accomplished purely by `ctags` if that's what you are suggesting.

**Reply by masatake** [reply]
You are correct. Parsing a Toml file is just a starting point. `-r` is just a vague idea.

**Reply by yancyribbens** [reply]
I don't think parsing a Toml file is helpful in this situation.  What I wrote never parses a Toml file but instead parses the JSON returned from cargo.

**Reply by masatake** [reply]
I will never think about running cargo from ctags. So we can expect the json output.

So ctags will do what cargo does about Cargo.toml partially.

I don't know both Rust and Cargo well. However, after I saw the output of run-ctags, 
I wrote an imitation:

~/var/ctags-github/cargo-list
```
#!/bin/bash

list_deps()
{
    local found;
    while read; do
	if [[ "$found" = 1 && "$REPLY" =~ ^([a-zA-Z0-9][^ \t=]*).* ]]; then
	    echo ${BASH_REMATCH[1]}
	fi
	
	if [[ "$REPLY" =~ ^\[dependencies\] ]]; then
	    found=1
	elif [[ "$REPLY" =~ ^\[ ]]; then
	    found=0
	    if [[ "$REPLY" =~ ^\[(dev-)?dependencies\.([^\]]*) ]]; then
		echo ${BASH_REMATCH[2]}
	    fi
	fi
    done < $1 | sort | uniq
}

get_cargos()
{
    local d
    local f
    
    for d in $@; do
	for f in ~/.cargo/registry/src/github.com-*/$d-*/Cargo.toml; do 
	    if [[ -f $f && -r $f ]]; then
		echo $f
	    fi
	done
     done | sort | uniq
}

list_deps_r()
{
    local r=$2
    
    if [[ $r > 3 ]]; then
	return
    fi
    
    local c
    local deps=$(list_deps $1)
    local cargos=$(get_cargos $deps)
    local deps_r=$({
		      for c in $cargos; do
			  list_deps $c
		      done
		  } | sort | uniq)
    {
	for c in ${cargos}; do
	    echo $c
	done

	r=$((r + 1))
	for c in $(get_cargos $deps $deps_r); do
	    list_deps_r $c $r
	done
    }  | sort | uniq
}

realpath $(dirname $1)
{
    list_deps_r $1 0 
} | xargs -n 1 dirname
```


```
$ git clone https://github.com/yancyribbens/run-ctags; cd run-ctags
$ bash ~/var/ctags-github/cargo-list ./Cargo.toml 
/home/jet/var/run-ctags
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/itoa-1.0.5
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro2-1.0.49
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/quote-1.0.23
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/ryu-1.0.12
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.151
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_derive-1.0.151
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_json-1.0.91
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.107
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-ident-1.0.6
$ run-ctags                                        
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/itoa-1.0.5
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro2-1.0.49
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/quote-1.0.23
/home/jet/var/run-ctags
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/ryu-1.0.12
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.151
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_derive-1.0.151
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_json-1.0.91
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.107
/home/jet/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-ident-1.0.6
```

I guess cargo does much more things.

Anyway, -r option may need the logic illustrated in the imitation.

I don't think about providing the feature to get the source code from repositories or somewhere to solve dependencies.
